### PR TITLE
classes for 4xx response body. start of close #117

### LINKED ISSUE #117: Error response body
Docker registry is expected to respond with certain error descriptions when problems occur. Right now error responses such as invalid digest, blob not found etc are served with empty bodies.
See docs fro proper error reporting fromat: https://docs.docker.com/registry/spec/api/#errors and https://docs.docker.com/registry/spec/api/#errors-2

Example of `UNAUTHROZIED` error response from `https://registry-1.docker.io/v2/`:
```
{"errors":[{"code":"UNAUTHORIZED","message":"authentication required","detail":null}]}
```

### PR DESCRIPTION
**Pull request by artemlazarev**
Classes for erros response. 
Start of close #117.

### CODE REVIEW

File: src/main/java/com/artipie/docker/http/ErrorResponse.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+     * Origin response.
+     */
+    private final Response origin;
+
+    /**
+     * Ctor.
+     * @param status Http status
+     * @param buff Origin response
+     */
+    ErrorResponse(final RsStatus status, final ByteBuffer buff) {
```
Comment (olegmoz) [inline src/main/java/com/artipie/docker/http/ErrorResponse.java:52]: @artemlazarev what use cases for this constructor do you see? I think that code using this class should not build body, this is responsibility of this class.

File: src/main/java/com/artipie/docker/http/ErrorResponse.java [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+        this.origin = new RsWithBody(
+            new RsWithHeaders(
+                new RsWithStatus(status),
+                new Headers.From("Content-Type", "application/json")
```
Comment (olegmoz) [inline src/main/java/com/artipie/docker/http/ErrorResponse.java:56]: @artemlazarev examples in the docs usually contain charset as well, for example `Content-Type: application/json; charset=utf-8`. I'd suggest we add that and encode response body using UTF-8 encoding

File: src/main/java/com/artipie/docker/http/ErrorResponse.java [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+import com.artipie.http.rs.RsWithStatus;
+import java.nio.ByteBuffer;
+import java.util.concurrent.CompletionStage;
+
+/**
+ * Docker response in case of errors.
+ *
+ * @since 0.2
+ */
+final class ErrorResponse implements Response {
```
[the remark this answers is not in the document]
Comment (artemlazarev) [reply src/main/java/com/artipie/docker/http/ErrorResponse.java]: @oleg done

File: src/main/java/com/artipie/docker/http/ErrorResponse.java [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+        this(
+            status,
+            ByteBuffer.wrap(
+                String.format(
+                    "{ \"errors:\" [{\r\n            \"code\":%s ,\r\n            \"message\": %s,\\r\\n            \"detail\": %s\r\n        }\r\n    ]\r\n}",
+                    code,
+                    message,
+                    detail
+                ).getBytes()
+            )
```
[the remark this answers is not in the document]
Comment (artemlazarev) [reply src/main/java/com/artipie/docker/http/ErrorResponse.java]: @olegmoz done

File: src/main/java/com/artipie/docker/http/ErrorResponse.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (artemlazarev) [reply src/main/java/com/artipie/docker/http/ErrorResponse.java]: @olegmoz if structure of code is good enough, i will do it.

File: src/main/java/com/artipie/docker/http/DockerErrors.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+import com.artipie.http.Response;
+import com.artipie.http.rs.RsStatus;
+import java.util.concurrent.CompletionStage;
+
+/**
+ * Docker response in case of errors.
+ *
+ * @since 0.2
+ */
+enum DockerErrors implements Response {
```
[the remark this answers is not in the document]
Comment (artemlazarev) [reply src/main/java/com/artipie/docker/http/DockerErrors.java]: @olegmoz done

File: src/main/java/com/artipie/docker/http/ErrorResponse.java (hunk 1 above)
Comment (artemlazarev) [reply src/main/java/com/artipie/docker/http/ErrorResponse.java]: @olegmoz i done with refactoring according you comment, so that's class has gone out of PR.

File: src/main/java/com/artipie/docker/http/DockerErrors.java [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+import com.artipie.http.rs.RsWithStatus;
+import java.nio.ByteBuffer;
+import javax.json.Json;
+
+/**
+ * Docker response in case of errors.
+ *
+ * @since 0.2
+ */
+abstract class DockerErrors extends  Response.Wrap {
```
Comment (olegmoz) [inline src/main/java/com/artipie/docker/http/DockerErrors.java:40]: @artemlazarev do not see why you changed the name, `ErrorResponse` was nice and did fit better for an object

File: src/main/java/com/artipie/docker/http/DockerErrors.java [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+     * @param status Http status
+     * @param buff Origin response
+     */
+    private DockerErrors(final RsStatus status, final ByteBuffer buff) {
+        super(
+            new RsWithBody(
+                new RsWithHeaders(
+                    new RsWithStatus(status),
+                    new Headers.From(
+                    "Content-Type", "application/json"
```
Comment (olegmoz) [inline src/main/java/com/artipie/docker/http/DockerErrors.java:53]: @artemlazarev please see my previous comments regarding encoding in `Content-Type` and lack of `Content-Length` header

File: src/main/java/com/artipie/docker/http/DockerErrors.java [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+        BlobUnknownError(final String detail) {
+            super(RsStatus.BAD_REQUEST, "BLOB_UNKNOWN", "blob unknown to registry", detail);
+        }
+    }
+
+    /**
+     * Internal error with json body.
+     * @since 0.2
+     */
+    final class NanifestInvalidError extends DockerErrors {
```
Comment (olegmoz) [inline src/main/java/com/artipie/docker/http/DockerErrors.java:106]: @artemlazarev may we make those errors not private classes? Referencing them as `SomethingError` will be shorter then `DockerErrors.SomethingError` and still quite clear. Don't you agree?

Comment (olegmoz) [inline src/main/java/com/artipie/docker/http/DockerErrors.java:106]: @artemlazarev there is typo in class name, should be `ManifestInvalidError` instead of `NanifestInvalidError`

File: src/main/java/com/artipie/docker/http/DockerErrors.java [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+                        .add("message", message)
+                        .add("detail", detail)
+                )
+                .build().toString().getBytes()
+            )
+        );
+    }
+
+    /**
+     * Internal error with json body.
```
Comment (olegmoz) [inline src/main/java/com/artipie/docker/http/DockerErrors.java:87]: @artemlazarev non informative class comment

File: src/main/java/com/artipie/docker/http/DockerErrors.java [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+         *
+         * @param detail Detail of the response.
+         */
+        BlobUnknownError(final String detail) {
+            super(RsStatus.BAD_REQUEST, "BLOB_UNKNOWN", "blob unknown to registry", detail);
+        }
+    }
+
+    /**
+     * Internal error with json body.
```
Comment (olegmoz) [inline src/main/java/com/artipie/docker/http/DockerErrors.java:103]: @artemlazarev non informative class comment

File: src/main/java/com/artipie/docker/http/ErrorResponse.java (hunk 2 above)
Comment (artemlazarev) [reply src/main/java/com/artipie/docker/http/ErrorResponse.java]: @olegmoz done

File: src/main/java/com/artipie/docker/http/DockerErrors.java (hunk 7 above)
Comment (artemlazarev) [reply src/main/java/com/artipie/docker/http/DockerErrors.java]: @olegmoz done

File: src/main/java/com/artipie/docker/http/DockerErrors.java (hunk 8 above)
Comment (artemlazarev) [reply src/main/java/com/artipie/docker/http/DockerErrors.java]: @olegmoz done

Comment (artemlazarev) [reply src/main/java/com/artipie/docker/http/DockerErrors.java]: @olegmoz done

File: src/main/java/com/artipie/docker/http/DockerErrors.java (hunk 10 above)
Comment (artemlazarev) [reply src/main/java/com/artipie/docker/http/DockerErrors.java]: @olegmoz done

**Comment by artemlazarev** [comment]
> @artemlazarev CI checks are failing, please fix them
@olegmoz sorry. Fogot to run quilice befor reqest of review

File: src/main/java/com/artipie/docker/http/ErrorResponse.java [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+import com.artipie.http.rs.RsWithStatus;
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+import javax.json.Json;
+
+/**
+ * Represent docker errors.
+ * @since 0.2
+ *
+ * Docker response in case of errors.
```
Comment (olegmoz) [inline src/main/java/com/artipie/docker/http/ErrorResponse.java:41]: @artemlazarev this string looks like some misplaced class description

File: src/test/java/com/artipie/docker/http/ErrorResponseTest.java [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+            new RsHasStatus(RsStatus.BAD_REQUEST)
+        );
+        MatcherAssert.assertThat(
+            "Does not put blob into storage",
+            this.storage.exists(
+                new BlobKey(new Digest.Sha256(content))
+            ).join(),
+            new IsEqual<>(false)
+        );
+    }
```
Comment (olegmoz) [inline src/test/java/com/artipie/docker/http/ErrorResponseTest.java:113]: @artemlazarev not sure what this test is testing. It involves `Storage`, `Docker` and `DockerSlice`. It does not tests `ErrorResponse` in any way. This test is suited to test particular `ErrorResponse` usage, which is not part of this PR

File: src/main/java/com/artipie/docker/http/ErrorResponse.java [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+    }
+
+    /**
+     * Ctor.
+     *
+     * @param status HTTP status of response
+     * @param code Code of response
+     * @param message Message of response
+     * @param detail More detail about of  response
+     * @checkstyle LineLengthCheck (10 lines)
```
Comment (olegmoz) [inline src/main/java/com/artipie/docker/http/ErrorResponse.java:76]: @artemlazarev seems to be pointless

File: src/main/java/com/artipie/docker/http/ErrorResponse.java [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+    ErrorResponse(final RsStatus status, final String code, final String message, final String detail) {
+        this(
+            status,
+            ByteBuffer.wrap(
+                Json.createObjectBuilder()
+                    .add(
+                        "errors",
+                        Json.createObjectBuilder().add("code", code)
+                            .add("message", message)
+                            .add("detail", detail)
```
Comment (olegmoz) [inline src/main/java/com/artipie/docker/http/ErrorResponse.java:88]: @artemlazarev sorry for not noticing this earlier, `errors` is supposed to be an array

File: src/main/java/com/artipie/docker/http/ErrorResponse.java [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+import com.artipie.http.rs.RsWithBody;
+import com.artipie.http.rs.RsWithHeaders;
+import com.artipie.http.rs.RsWithStatus;
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+import javax.json.Json;
+
+/**
+ * Represent docker errors.
+ * @since 0.2
```
Comment (olegmoz) [inline src/main/java/com/artipie/docker/http/ErrorResponse.java:39]: @artemlazarev should be since 0.3

File: src/main/java/com/artipie/docker/http/DigestInvalidError.java [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+ * SOFTWARE.
+ */
+package com.artipie.docker.http;
+
+import com.artipie.http.rs.RsStatus;
+
+/**
+ * DIGEST_INVALID error message with all json fields at the response body.
+ *
+ * @since 0.2
```
Comment (olegmoz) [inline src/main/java/com/artipie/docker/http/DigestInvalidError.java:31]: @artemlazarev should be since 0.3

File: src/main/java/com/artipie/docker/http/DockerErrors.java (hunk 6 above)
Comment (artemlazarev) [reply src/main/java/com/artipie/docker/http/DockerErrors.java]: @olegmoz done

File: src/main/java/com/artipie/docker/http/DockerErrors.java (hunk 9 above)
Comment (artemlazarev) [reply src/main/java/com/artipie/docker/http/DockerErrors.java]: @olegmoz  outdated.  fixed already

File: src/main/java/com/artipie/docker/http/ErrorResponse.java [hunk 17]
```diff
@@ -0,0 +1,94 @@
...
+     * @param status Http status
+     * @param buff Origin response
+     */
+    private ErrorResponse(final RsStatus status, final ByteBuffer buff) {
+        super(
+            new RsWithBody(
+                new RsWithHeaders(
+                    new RsWithStatus(status),
+                    new Headers.From(
+                        new Header("Content-Length", String.valueOf(buff.position())),
```
[the remark this answers is not in the document]
Comment (artemlazarev) [reply src/main/java/com/artipie/docker/http/ErrorResponse.java:60]: @olegmoz  So you think get size of buffer with text and pass it from constructor will be better?

File: src/main/java/com/artipie/docker/http/ErrorResponse.java (hunk 11 above)
Comment (artemlazarev) [reply src/main/java/com/artipie/docker/http/ErrorResponse.java:41]: @olegmoz  done

File: src/main/java/com/artipie/docker/http/ErrorResponse.java (hunk 13 above)
Comment (artemlazarev) [reply src/main/java/com/artipie/docker/http/ErrorResponse.java:76]: @olegmoz qulice :)

Comment (olegmoz) [reply src/main/java/com/artipie/docker/http/ErrorResponse.java:76]: @artemlazarev if it is complaining about line with parameters then it could be broken into multiple lines. It is better then just suppressing it

File: src/main/java/com/artipie/docker/http/ErrorResponse.java (hunk 17 above)
[the remark this answers is not in the document]
Comment (olegmoz) [reply src/main/java/com/artipie/docker/http/ErrorResponse.java:60]: @artemlazarev I think it would be better to take size from byte array. Also there no point in having private constructor, it is used only once anyway. All the code could be in one place
